### ISSUE TITLE
Company hangs Emacs in Geiser REPL

### ISSUE THREAD

**Issue by raxod502** [jaor/geiser#183]
Steps to reproduce:
- `M-x run-racket`.
- In the Geiser REPL, `(require xrepl)`.
- Start typing `(values 1 2)`.

In my setup (on an OS X virtual machine), this causes Emacs to hang after typing `(v`. If I'm in a terminal Emacs, that's it—I have to restart Emacs. However, if I'm in a windowed Emacs (`emacsw`), I can regain control by pressing `C-g C-g C-g`.

Turning off `company-mode` prevents the freeze from occurring.

This also happens occasionally in other, non-`xrepl` situations. However, this case is the most reproducible one I have found.

Following the instructions in [this Stack Exchange question](http://emacs.stackexchange.com/a/507/12534), I produced the following stack trace:

```
Debugger entered--Lisp error: (quit)
  accept-process-output(#<process Racket REPL> 3.0)
  byte-code("\305\306 \236A!)\205'\n\307\n\310\311 \"A\"*?\205'\312\f\313\245\"\210\202\207" [con c id proc timeout tq-process :tq gethash assoc :completed accept-process-output 10] 5)
  byte-code("\303\304\305#\211\nB\306\216\304\307\310\217+\207" [timeout -with-timeout-timer- with-timeout-timers run-with-timer nil #[nil "\300\301\211\"\207" [throw timeout] 3] ((cancel-timer -with-timeout-timer-)) (byte-code "\305\306  \236A!)\205'\n\307\n\310\311 \"A\"*?\205'\312\f\313\245\"\210\202\207" [con c id proc timeout tq-process :tq gethash assoc :completed accept-process-output 10] 5) ((error (byte-code "\301\236\302\241)\207" [req :continuation nil] 2)))] 5)
  geiser-con--send-string/wait((t (:filter . comint-output-filter) (:tq ((nil "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 82) (:string . ",geiser-eval #f #f ,apply geiser:completions (\"s\")") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer * Racket REPL *>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 83) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"v\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 83) (:string . ",geiser-eval #f #f ,apply geiser:completions (\"v\")") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer * Racket REPL *>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 83) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request)) #<process Racket REPL> . #<buffer  tq-temp-Racket REPL-153340>) (:tq-filter lambda (p s) (geiser-con--tq-filter (quote (((nil "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"v\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request)) #<process Racket REPL> . #<buffer  tq-temp-Racket REPL-153340>)) s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 83) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...)))) ",geiser-eval #f #f ,apply geiser:completions (\"v\")" geiser-eval--set-sync-retort nil nil)
  geiser-eval--send/wait((:eval (:ge completions "v")) nil nil)
  geiser-completion--symbol-list("v")
  geiser-completion--complete("v" nil)
  geiser-completion--thing-at-point(nil nil)
  geiser-completion--for-symbol()
  completion--capf-wrapper(geiser-completion--for-symbol optimist)
  run-hook-wrapped(completion--capf-wrapper geiser-completion--for-symbol optimist)
  company--capf-data-real()
  company--capf-data()
  company-capf(prefix)
  #[0 "\n\203   \nC\202\n\211\205\233\211@\3009\204\306!\203<9\2036\307N\310=\2046\307N?\205@\311!\205@\312!\202@\313\312\"\240\210\300\242\203\224\314\300\242!\203\216\315\300\242!\211\301\316\f!\240\210\301\242:\204w
\203i\317\320!\210\301\242\310=\203\216\321\322!\210\202\216
\203\300\242\323\301\242!\210\324\325\326 \"\210\327\330!\210\331\332\301\242\"\210A\266\202\202\n\207" [(nil) (nil) company-backend company-backends company-prefix company--manual-action functionp company-init t company-init-backend prefix company--multi-backend-adapter company--good-prefix-p company--prefix-str company-calculate-candidates message "No completion found" company-cancel unique company-update-candidates run-hook-with-args company-completion-started-hook company-explicit-action-p company-call-frontends show throw --cl-block-nil-- company--manual-prefix] 6 "\n\n(fn)"]()
  funcall(#[0 "\n\203   \nC\202\n\211\205\233\211@\3009\204\306!\203<9\2036\307N\310=\2046\307N?\205@\311!\205@\312!\202@\313\312\"\240\210\300\242\203\224\314\300\242!\203\216\315\300\242!\211\301\316\f!\240\210\301\242:\204w
\203i\317\320!\210\301\242\310=\203\216\321\322!\210\202\216
\203\300\242\323\301\242!\210\324\325\326 \"\210\327\330!\210\331\332\301\242\"\210A\266\202\202\n\207" [(nil) (nil) company-backend company-backends company-prefix company--manual-action functionp company-init t company-init-backend prefix company--multi-backend-adapter company--good-prefix-p company--prefix-str company-calculate-candidates message "No completion found" company-cancel unique company-update-candidates run-hook-with-args company-completion-started-hook company-explicit-action-p company-call-frontends show throw --cl-block-nil-- company--manual-prefix] 6 "\n\n(fn)"])
  company--begin-new()
  company--perform()
  #[0 "\301 \210\207" [company-candidates company--perform] 1 "\n\n(fn)"]()
  funcall(#[0 "\301 \210\207" [company-candidates company--perform] 1 "\n\n(fn)"])
  company-auto-begin()
  company-idle-begin(#<buffer * Racket REPL *> #<window 26 on * Racket REPL *> 193 55)
  apply(company-idle-begin (#<buffer * Racket REPL *> #<window 26 on * Racket REPL *> 193 55))
  byte-code("r\301\302H\303H\"\210)\301\207" [timer apply 5 6] 4)
  timer-event-handler([t 22510 65082 123683 nil company-idle-begin (#<buffer * Racket REPL *> #<window 26 on * Racket REPL *> 193 55) nil 0])
```

Likely this is due to the Geiser or Racket functions called to look up completions hanging (which causes Company to hang).

**Comment by jaor** [comment]
i'll take a look, but, in general, supporting xrepl is quite hairy and
something i cannot commit to: it modifies the workings of the repl too
deeply to be really compatible with geiser's repl... geiser also defines
its own read-eval-print loop for racket, so there are lots of
opportunities for compatibility problems between the too.

on the other hand, if the hangs happen without require xrepl, that's
definitely a bug we want to fix.

**Comment by raxod502** [comment]
OK. I will see if I can find a reproducible test case that does not involve `xrepl`.

Actually, the only thing I wanted `xrepl` for was to be able to use previous evaluation results (`^`, `^^`, `^^^`…). Does Geiser provide this functionality separately?

**Comment by jaor** [comment]
On Sat, Oct 01 2016, Radon Rosborough wrote:

> Actually, the only thing I wanted xrepl for was to be able to use
> previous evaluation results (^, ^^, ^^^…). Does Geiser provide this
> functionality separately?

not for racket.  it's available in guile, because the latter comes with
it built in.  but since geiser implements its own repl in racket, we
probably could add it... perhaps we could borrow guile's notation (`$1`,
`$2`, ...) to make it more homogeneous...

**Comment by raxod502** [comment]
I think that is an absolutely great idea! To be honest, I don't really understand why this functionality isn't included to begin with.

**Comment by turbopape** [comment]
Hey guys,
I have the same problem with other schemes - chibi - and I am not using xRepl.
company when used for cpmpletion freezes, and it's only after doing C-G C-G that I gain control over my Emacs again.
i am on mac OSX, my Emacs version is:
GNU Emacs 25.1.1 (x86_64-apple-darwin13.4.0, NS appkit-1265.21 Version 10.9.5 (Build 13F1911)) of 2016-09-17

**Comment by ylluminarious** [comment]
@jaor Same issue is occurring for me also; it seems to happen when I import almost any module or file via `C-c C-m` in the REPL or `C-c C-a` in a corresponding file buffer. I'm on OS X 10.12.3, GNU Emacs 25.1.1.

**Comment by jaor** [comment]
@ylluminarious thanks for the report... it's hard for me to test because i don't have a mac, and i'm not seeing the problem on gnu/linux.  is this something that also happens without ac-geiser installed?

**Comment by ylluminarious** [comment]
@jaor Thanks for the response. Yes, looking at these other comments here, the common thread between them appears to be OS X so I imagine that it must have something to do with what's going wrong. So, I went ahead and uninstalled `ac-geiser` as per your request and have found that the issue does not occur once `ac-geiser` no longer installed and rudimentary auto-completion will work for symbols that have already been typed in the buffer. So this evidently something that is being triggered specifically by both `ac-geiser` and `company-mode`.

**Comment by ylluminarious** [comment]
@jaor Were you able to ascertain anything useful from my above comment?

**Comment by jaor** [comment]
not yet, but i haven't had time to look at things with care... probably
during the weekend!

**Comment by ylluminarious** [comment]
@jaor Okay, feel free to ping me if you need further information.

**Comment by ylluminarious** [comment]
@jaor Any progress on this? No rush, just curious.

**Comment by jaor** [comment]
On Sat, Feb 18 2017, George D. Plymale II wrote:

> @jaor Any progress on this? No rush, just curious.

sorry, not yet... i haven't managed to get a free slot yet, but please
bear with me :)

**Comment by ylluminarious** [comment]
@jaor No problem, feel free to ask for more info if needed.

**Comment by ylluminarious** [comment]
@jaor Any updates on this? No rush, just wondering if this will be solved anytime soon.

**Comment by ylluminarious** [comment]
@jaor FYI, I was testing this out recently and it appears that running the `completion-at-point` command will also cause this hanging problem. So it seems that pretty much *any* completion that geiser tries to perform after loading a module / file causes it to choke and hang the process.

**Comment by jaor** [comment]
George, does C-u M-x geiser-show-logs and then reproducing the problem
produce any traces in the log window?

On 8 Mar 2017 00:37, "George D. Plymale II" <notifications@github.com>
wrote:

@jaor <https://github.com/jaor> FYI, I was testing this out recently and it
appears that running the completion-at-point command will also cause this
hanging problem. So it seems that pretty much *any* completion that geiser
tries to perform after loading a module / file causes it to choke and hang
the process.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/jaor/geiser/issues/183#issuecomment-284896704>, or mute
the thread

**Comment by ylluminarious** [comment]
@jaor Ok, so I entered the `swindle` module via `C-c C-m swindle RET` and then typed `(au)` (for no reason in particular)  in the REPL prompt like so:

```
Welcome to Racket v6.8.
racket@> ,enter swindle
=racket@main.rkt> (au)
```

Emacs hung, I did `$ kill -SIGUSR2 <emacs_pid>` and this is what was in the log buffer:

```

INFO: REQUEST: <1>: ,geiser-eval geiser/user #f (image-cache)

INFO: RETORT: ((result "\"/var/folders/zs/22kr6y0x0114jt6ly1s5jj7w0000gq/T/\"") (output . ""))

INFO: <1>: processed

INFO: REQUEST: <2>: ,geiser-eval #f #f ,apply geiser:completions ("au")

INFO: REQUEST: <3>: ,geiser-eval #f #f ,apply geiser:autodoc ('(au au))

INFO: REQUEST: <4>: ,geiser-eval #f #f ,apply geiser:completions ("au")

INFO: REQUEST: <5>: ,geiser-eval #f #f ,apply geiser:completions ("au")

```

**Comment by jaor** [comment]
hmmm, does disabling autodoc in the repl help in any way? (as an aside,
swindle is a tricky module that might break geiser in other ways, but i
take this is not module specific)

**Comment by ylluminarious** [comment]
Correct, this not module-specific (as far as I know), but I'll keep that in mind about `swindle` . When I try entering `xrepl` instead (since that was the original topic of this ticket), the same thing happens even after disabling autodoc in the REPL.

**Comment by jaor** [comment]
thanks. I'll try to reproduce it on my side again (but i must say, xrepl is
even more prone than swindle to break geiser :))

**Comment by ylluminarious** [comment]
@jaor Ok, thanks. Sorry for the quirky choice of modules :P ; I'll test other modules if necessary.

**Comment by ylluminarious** [comment]
@jaor Hey, I don't mean to bother you, but I was reminded of this issue again the other day and wanted to inform you that  when I entered the `typed/racket` module and  used code completion in the REPL, it did not hang and things seemed to work normally. I'm not sure why this is, but it suggests that there is in fact something module-specific about this issue.

**Comment by jaor** [comment]
thanks for the update... this is indeed funny, typed-racket is about the
most troublesome i can think of :)

**Comment by ylluminarious** [comment]
@jaor Huh, wow, ok. Do let me know if you need more info, testing, or whatnot.

**Comment by Saikyun** [comment]
This started happening to me in geiser using guile. I don't know why it started, but now whenever I try to complete a variable or evaluate an expression emacs just stops.

**Comment by ylluminarious** [comment]
@Saikyun Does that happen when you enter a module or file? Or does that just happen right off the bat when trying to complete anything in a totally vanilla REPL? The common thread here thus far has been entering a module or file, but if that's not what's causing the issue I imagine that will be helpful for @jaor to know to fix the problem.

**Comment by jaor** [comment]
update on this: i just got access to a mac, so i'll hopefully be able to
reproduce it on my side.  still short of time around here, so please
keep being that incredibly kind with me! :)

**Comment by ylluminarious** [comment]
@jaor Great, thanks for update.

**Comment by ylluminarious** [comment]
@Saikyun Okay, thanks.

**Comment by Saikyun** [comment]
@ylluminarious To actually answer your question; it happened when I either evaluated a line (C-x C-e) or autocompleted by pressing tab by using these settings:
```(setq tab-always-indent 'complete)
(add-to-list 'completion-styles 'initials t)```

But since I was in the middle of debugging and I autoload a bunch of stuff it's hard to say what it actually tried to do.

**Comment by ylluminarious** [comment]
@Saikyun Thanks for the info. The part about using `C-x C-e` to trigger the hang is particularly interesting because it doesn't have to do with completion. All I had to do to reproduce it on my end was:

1. Start the REPL and enter `swindle` (because I already know that one's problematic)
2. Then I went into a Scheme buffer and tried to evaluate `(define x 1)` via `C-x C-e`

Then, this was in the Geiser log buffer after breaking the hang via `$ kill -SIGUSR2 <emacs_pid>`:
```
INFO: REQUEST: <9>: ,geiser-eval #f #f (define x 1)
```
And this was in the `*Backtrace*` buffer:
```
Debugger entered--Lisp error: (quit)
  accept-process-output(#<process Racket REPL> 3.0)
  geiser-con--send-string/wait((t (:filter . comint-output-filter) (:tq ((nil "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 2) (:string . ",geiser-eval #f #f ,apply geiser:autodoc ('(define))") (:continuation . geiser-autodoc--show-signatures) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(define))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 3) (:string . ",geiser-eval #f #f ,apply geiser:autodoc ('(define))") (:continuation . geiser-autodoc--show-signatures) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(1 define))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 4) (:string . ",geiser-eval #f #f ,apply geiser:autodoc ('(1 define))") (:continuation . geiser-autodoc--show-signatures) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 5) (:string . ",geiser-eval #f #f (define x 1)") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"de\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 6) (:string . ",geiser-eval #f #f ,apply geiser:completions (\"de\")") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer * Racket REPL *>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(de de))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 7) (:string . ",geiser-eval #f #f ,apply geiser:autodoc ('(de de))") (:continuation . geiser-autodoc--show-signatures) (:buffer . #<buffer * Racket REPL *>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"de\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 8) (:string . ",geiser-eval #f #f ,apply geiser:completions (\"de\")") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer * Racket REPL *>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 9) (:string . ",geiser-eval #f #f (define x 1)") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" ((:id . 10) (:string . ",geiser-eval #f #f (define x 1)") (:continuation . geiser-eval--set-sync-retort) (:buffer . #<buffer tmp>) (:connection t (:filter . comint-output-filter) #1 (:tq-filter lambda (p s) (geiser-con--tq-filter ... s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...))))) . geiser-con--process-completed-request)) #<process Racket REPL> . #<buffer  tq-temp-Racket REPL>) (:tq-filter lambda (p s) (geiser-con--tq-filter (quote (((nil "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(define))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(1 define))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"de\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:autodoc ('(de de))\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f ,apply geiser:completions (\"de\")\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request) (",geiser-eval #f #f (define x 1)\n" "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)" (... ... ... ... ...) . geiser-con--process-completed-request)) #<process Racket REPL> . #<buffer  tq-temp-Racket REPL>)) s)) (:eot . "\\(\n\\(mzscheme\\|racket\\)@[^ ]*> \\)") (:prompt . "\\(mzscheme\\|racket\\)@[^ ]*> ") (:debug-prompt) (:is-debugging) (:count . 10) (:completed . #s(hash-table size 65 test eql weakness value rehash-size 1.5 rehash-threshold 0.8 data ( ...)))) ",geiser-eval #f #f (define x 1)" geiser-eval--set-sync-retort nil nil)
  geiser-eval--send/wait((:eval (:scm "(define x 1)")))
  geiser-debug--send-region(nil 1 13 nil nil nil)
  geiser-eval-region(1 13 nil t nil)
  geiser-eval-last-sexp(nil)
  funcall-interactively(geiser-eval-last-sexp nil)
  #<subr call-interactively>(geiser-eval-last-sexp nil nil)
  ad-Advice-call-interactively(#<subr call-interactively> geiser-eval-last-sexp nil nil)
  apply(ad-Advice-call-interactively #<subr call-interactively> (geiser-eval-last-sexp nil nil))
  call-interactively(geiser-eval-last-sexp nil nil)
  command-execute(geiser-eval-last-sexp)
```

@jaor Any thoughts on this now? It seems that the core of this problem may not be completion at all.

**Comment by jaor** [comment]
On Tue, May 16 2017, George D. Plymale II wrote:

[...]

> @jaor Any thoughts on this now? It seems that the core of this problem
> may not be completion at all.

my first impression was that the problem is specific to swindle... that
module redefines lots of low-level stuff in the language and it's likely
that in the process it alters some of the evaluation semantics that the
racket side of geiser relies upon (there's similar problems almost
everytime we try geiser with non-standard #lang's).

i'd expect the same kind of trouble from a module such as xrepl, which
again tweaks racket's evaluation and repl.

were you able to reproduce these problems entering less heavy modules?
(sorry, i know i must have asked this before...)

**Comment by ylluminarious** [comment]
@jaor So far (to my recollection and upon re-reading this thread) I've only tested `xrepl` (which does trigger the bug) and I've also tested `typed/racket` (which strangely does not trigger the bug). I can test other weird (and less heavy) modules if you'd like.

**Comment by jaor** [comment]
On Mon, May 29 2017, George D. Plymale II wrote:

> @jaor So far (to my recollection and upon re-reading this thread) I've
> only tested xrepl (which does trigger the bug) and I've also tested
> typed/racket (which strangely does not trigger the bug). I can test
> other weird (and less heavy) modules if you'd like.

that could be useful, if you have the bandwidth.

but if the problem only happens with modules such as xrepl that define
their own repls, it's going to be difficult to fix, since each module
will probably require an ad-hoc solution, and geiser would need to
become aware of specific modules to adapt its behaviour.  i don't see
any really viable solution in general to that.

specifically, supporting xrepl is a non-goal, since geiser and xrepl are
implementations of the same thing in different ways: in my opinion, what
we should do there is listing the features of xrepl that we'd like to
see in geiser and reimplement them (preferably for all supported
schemes, not just racket; although typically that'd mean for guile).

supporting swindle would be nice, but sounds very hard, and i know very
few racketeers (or racket libraries) using it, so it'd probably be low
in my priority list...

but, if the problem is reproducible in user-created modules, or in fact
any `#lang racket lib`, it should be fixed with high priority.

**Comment by ylluminarious** [comment]
@marsam I tested your theory with the modules that I was having problems with and I noticed that they all do indeed change the prompt. Moreover, all the modules that worked fine do not change the prompt. However, when I tried adjusting `geiser-racket--prompt-regexp` to match the new prompt, Emacs would still hang... am I missing something? Or, perhaps, is this not the root cause of the issue?

@jaor Please chime in if you have any thoughts on this.

**Comment by jaor** [comment]
george, i think it is very plausible that changing the prompt is at the
root of the problem, or at least very near its root :).  how did you
change geiser-racket--prompt-regexp?

**Comment by ylluminarious** [comment]
@jaor I ran the following in `*ielm*`:
```lisp
(setq geiser-racket--prompt-regexp "<pkgs>/xrepl-lib/xrepl/main> ")
```

**Comment by jaor** [comment]
On Wed, Sep 27 2017, George D. Plymale II wrote:

> @jaor I ran the following in *ielm*:
>
> (setq geiser-racket--prompt-regexp "<pkgs>/xrepl-lib/xrepl/main> ")

it's probably a good idea to also run `(geiser-reload)` after that (or
`M-x geiser-reload`), so that the full geiser system notices the
chance... but, setting that hard-coded prompt is going to make geiser
only functional when you are inside that module, so i don't think that
is going to fly: to enter the module, you'll be first in an outter one,
without a functional geiser, so you won't be able to enter it.

you should come up with a regular expression that encompasses both
cases...

**Comment by ylluminarious** [comment]
@jaor Thanks for the advice. I ran the following (the new regexp encompasses both prompt styles):
```lisp
(setq geiser-racket--prompt-regexp "<pkgs>.*> \\|\\(mzscheme\\|racket\\)@[^ ]*> ")
```
And now Geiser no longer hangs when I try to complete symbols in the REPL (yay!). The only quirk now is that auto-completion doesn't seem to happen at all, but I suppose that falls outside the scope of this ticket.

Can we mark this ticket as closed or should there be some modification to Geiser's source code to fix this issue? The only problem that I see with fixing it is that we can't possibly make a regexp to encompass all possible prompt styles... as an alternative, we could just make a notice in the documentation which states how to work around this issue if one should encounter it.

**Comment by jaor** [comment]
excellent. thanks for the update.  yes, i think a notice, possibly
including xrepl as an example, is a good compromise here.

**Comment by ylluminarious** [comment]
@jaor Alright, I'll try to write something up and submit it as a patch today or tomorrow. Or you can write it up if you'd like; doesn't matter to me.

**Comment by ylluminarious** [comment]
Also thanks a lot @marsam, for leading us to the solution of this issue!

**Comment by ylluminarious** [comment]
@jaor Please see https://github.com/jaor/geiser/pull/223 for my patch as we discussed above.

**Comment by ylluminarious** [comment]
@jaor With the merging of #223, should this be marked as closed?

**Comment by jaor** [comment]
yes, i think so.  if any of the participants above feels otherwise, please feel free to reopen it.
